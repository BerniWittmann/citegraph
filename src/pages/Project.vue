<template>
  <div class="project">
    <v-alert v-if="!hasProject" type="error" prominent>
      {{ $t('project.not_available') }}
    </v-alert>

    <div v-if="hasProject">
      <h1>{{ project.name }}</h1>
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator'
import Project from '@/models/project'

@Component
export default class ProjectPage extends Vue {
  get project (): Project | undefined {
    return this.$store.getters['projects/activeProject']
  }

  get hasProject (): boolean {
    return this.$store.getters['projects/hasActiveProject']
  }
}
</script>
