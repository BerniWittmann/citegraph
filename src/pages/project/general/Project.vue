import { ProjectStates } from '@/models/project'
<template>
  <div class="project">
    <div class="d-flex mr-5 mt-4">
      <h1>{{ project.name }}</h1>
      <v-spacer/>
      <v-btn :to="{ name: 'projects.single.edit', params: { projectId: project.id } }">
        <v-icon>mdi-pencil</v-icon>
        {{ $t('project.edit.button_text') }}
      </v-btn>
    </div>
    <v-container v-if="hasNoDataYet">
      <v-row justify="center">
        <v-col :lg="3" class="d-flex justify-center flex-column text-center">
          <empty-icon class="align-self-center mb-6"></empty-icon>
          <h3 class="headline">{{ $t('project.single.no_data.description') }}</h3>
          <p>{{ $t('project.single.no_data.import_data') }}</p>
          <v-btn color="primary" rounded large :to="{ name: 'projects.single.import', params: { projectId: project.id } }">
            {{ $t('project.single.no_data.button') }}
          </v-btn>
        </v-col>
      </v-row>
    </v-container>
    <v-container v-else>
      <v-row v-if="datasetInformation">
        <v-col>
          <v-card>
            <v-card-title>{{ $t('project.single.dataset.title') }}</v-card-title>
            <v-card-text>
              <v-container>
                <v-row>
                  <v-col v-for="(count, name) in datasetInformation" :key="name">
                    <v-card outlined class="d-flex flex-column justify-center text-center pa-5">
                      <p class="headline">{{count}}</p>
                      <p class="mb-0">{{ $t('project.single.dataset.' + name ) }}</p>
                    </v-card>
                  </v-col>
                </v-row>
              </v-container>
            </v-card-text>
          </v-card>
        </v-col>
      </v-row>
      <v-row>
        <v-col lg="3" md="6" cols="12" v-if="project.topAuthors">
          <v-card>
            <v-card-title>
              {{ $t('project.single.overview.top_10_authors') }}
            </v-card-title>
            <v-card-text>
              <v-list>
                <v-list-item v-for="author in project.topAuthors" :key="author.id" :to="{ name: 'projects.single.explore.view', params: {
                 projectId: project.id,
                 queryByType: 'author',
                 entityId: author.id
                }}">
                  {{ author.firstName }} {{ author.lastName }}
                </v-list-item>
              </v-list>
            </v-card-text>
          </v-card>
        </v-col>
        <v-col lg="3" md="6" cols="12" v-if="project.topRecords">
          <v-card>
            <v-card-title>
              {{ $t('project.single.overview.top_10_records') }}
            </v-card-title>
            <v-card-text>
              <v-list>
                <v-list-item v-for="record in project.topRecords" :key="record.id" :to="{ name: 'projects.single.explore.view', params: {
                 projectId: project.id,
                 queryByType: 'record',
                 entityId: record.id
                }}">
                  {{ record.title }}
                </v-list-item>
              </v-list>
            </v-card-text>
          </v-card>
        </v-col>
        <v-col lg="3" md="6" cols="12" v-if="project.countryRelevanceData">
          <v-card>
            <v-card-title>
              {{ $t('project.single.overview.country_relevance') }}
            </v-card-title>
            <v-card-text>
              TODO Diagram here
            </v-card-text>
          </v-card>
        </v-col>
        <v-col lg="3" md="6" cols="12">
          <v-row no-gutters>
            <v-col v-if="project.topKeywords" class="mb-6">
              <v-card>
                <v-card-title>
                  {{ $t('project.single.overview.top_10_keywords') }}
                </v-card-title>
                <v-card-text>
                  <v-chip-group multiple column>
                    <v-chip v-for="keyword in project.topKeywords" :key="keyword">{{ keyword }}</v-chip>
                  </v-chip-group>
                </v-card-text>
              </v-card>
            </v-col>
          </v-row>
          <v-row no-gutters>
            <v-col v-if="project.counts">
              <v-card>
                <v-card-title>
                  {{ $t('project.single.overview.visualizations') }}
                </v-card-title>
                <v-card-text>
                  <div class="d-flex flex-column justify-center text-center pa-5" v-if="hasNoVisualizationsYet">
                    <p class="mb-0">{{ $t('project.single.overview.no_visualizations_yet') }}</p>
                    <v-btn color="primary" rounded>{{ $t('project.single.overview.add_visualization') }}</v-btn>
                  </div>
                  <v-card outlined class="d-flex flex-column justify-center text-center pa-5" v-else>
                    <p class="headline">{{ project.counts.visualizations }}</p>
                    <p class="mb-0">{{ $t('project.single.overview.visualizations') }}</p>
                  </v-card>
                </v-card-text>
              </v-card>
            </v-col>
          </v-row>
        </v-col>
      </v-row>
    </v-container>
  </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator'
import Project, { ProjectStates } from '@/models/project'
import EmptyIcon from '@/assets/svgs/Empty.vue'

@Component({
  components: {
    EmptyIcon
  }
})
export default class ProjectPage extends Vue {
  get project (): Project {
    return this.$store.getters['projects/activeProject']
  }

  get datasetInformation (): Object | undefined {
    if (!this.project.counts) return undefined
    const data = { ...this.project.counts }
    delete data['visualizations']
    return data
  }

  get hasNoDataYet (): boolean {
    return this.project.state === ProjectStates.IMPORT
  }

  get hasNoVisualizationsYet (): boolean {
    return this.project.state === ProjectStates.EXPLORE
  }
}
</script>
