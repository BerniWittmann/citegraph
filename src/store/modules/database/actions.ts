import { ActionTree } from 'vuex'
import { RootState } from '@/store/types'
import { DatabaseState } from '@/store/modules/database/types'
import Axios from 'axios'

export const actions: ActionTree<DatabaseState, RootState> = {
  async clearDatabase ({ dispatch }): Promise<undefined> {
    try {
      await Axios(`/database/clear`, {
        method: 'POST'
      })
      dispatch('toasts/showSuccess', 'settings.database.clear.success', { root: true })
      return undefined
    } catch (error) {
      console.error(error)
      dispatch('toasts/showError', 'settings.database.clear.error', { root: true })
      return error
    }
  },

  async reseedDatabase ({ dispatch }): Promise<undefined> {
    try {
      await Axios(`/database/reseed`, {
        method: 'POST'
      })
      dispatch('toasts/showSuccess', 'settings.database.reseed.success', { root: true })
      return undefined
    } catch (error) {
      console.error(error)
      dispatch('toasts/showError', 'settings.database.reseed.error', { root: true })
      return error
    }
  }
}
