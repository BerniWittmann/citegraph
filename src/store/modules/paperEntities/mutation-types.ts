export const SET_ENTITIES = 'SET_ENTITIES'
export const SET_ENTITY_TYPE = 'SET_ENTITY_TYPE'
export const SET_ENTITY_COUNT = 'SET_ENTITY_COUNT'
