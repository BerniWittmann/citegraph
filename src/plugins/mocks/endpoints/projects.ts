import { AxiosRequestConfig } from 'axios'
import MockAdapter from 'axios-mock-adapter'
import db from '../db'
import Project from '@/models/project'

const projects = db.getCollection('projects')

function getProjectId (config: AxiosRequestConfig): number {
  const urlParts: Array<string> = config.url!.split('/')
  return parseInt(urlParts[4])
}

export function setupMocks (mock: MockAdapter) {
  mock.onGet(/\/projects\/\d+/).reply((config: AxiosRequestConfig) => {
    const project: Resultset<Project> = projects.findOne({ 'id': getProjectId(config) })
    if (!project) {
      return [404]
    }
    return [200, project]
  })

  mock.onGet('/projects').reply(() => {
    return [200, projects.find()]
  })

  mock.onPost('/projects').reply((config: AxiosRequestConfig) => {
    try {
      const projectName = JSON.parse(config.data).name.trim()
      const newProject = new Project({ id: projects.count() + 1, name: projectName })
      projects.insert(newProject)
      db.saveDatabase()
      return [201, newProject]
    } catch {
      return [400, 'Could not read body']
    }
  })

  mock.onDelete(/\/projects\/\d+/).reply((config: AxiosRequestConfig) => {
    const project: Resultset<Project> = projects.findOne({ 'id': getProjectId(config) })
    if (!project) {
      return [404]
    }
    projects.remove(project)
    db.saveDatabase()
    return [200, project]
  })

  mock.onPut(/\/projects\/\d+/).reply((config: AxiosRequestConfig) => {
    let project: Project = projects.findOne({ 'id': getProjectId(config) })
    if (!project) {
      return [404]
    }
    const projectName = JSON.parse(config.data).name.trim()
    project = projects.update({
      ...project,
      name: projectName
    })
    db.saveDatabase()
    return [200, project]
  })
}
