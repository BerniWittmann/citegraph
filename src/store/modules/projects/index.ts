import { Module } from 'vuex'
import { ProjectsState } from './types'
import { RootState } from '@/store/types'
import { state } from './state'
import { actions } from './actions'
import { getters } from './getters'
import { mutations } from './mutations'

const namespaced: boolean = true

export const projects: Module<ProjectsState, RootState> = {
  namespaced,
  state,
  getters,
  actions,
  mutations
}
