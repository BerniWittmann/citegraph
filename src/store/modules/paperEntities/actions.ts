import { ActionTree } from 'vuex'
import { RootState } from '@/store/types'
import { PaperEntitiesState } from './types'
import { PaperEntityQueryParameters } from '@/models/paperEntities/query'
import Axios from 'axios'
import * as types from '@/store/modules/paperEntities/mutation-types'
import jsLogger from 'js-logger'
import { entityKeysMap, PaperEntity, PaperEntityFields } from '@/models/paperEntities'
import { PaperEntityMutationParameters } from '@/models/paperEntities/mutation'

const logger = jsLogger.get('paperEntities/actions')

export const actions: ActionTree<PaperEntitiesState, RootState> = {
  async fetchEntities ({ commit, dispatch }, params: PaperEntityQueryParameters): Promise<undefined> {
    try {
      const EntityClass = entityKeysMap[params.entityType]
      const response = await Axios(`/projects/${params.projectId}/paper-entities`, {
        method: 'POST',
        data: {
          query: EntityClass.getQuery(params)
        }
      })
      const data = response.data.data[EntityClass.queryName]
      const totalCount = data.count
      const payload: Array<PaperEntity> = EntityClass.transformer.fetchCollection(data[EntityClass.schemaName])
      commit(types.SET_ENTITIES, payload)
      commit(types.SET_ENTITY_COUNT, totalCount)
      commit(types.SET_ENTITY_TYPE, EntityClass.key)
      return undefined
    } catch (error) {
      logger.error(error)
      dispatch('toasts/showError', 'project.explore.fetch_error', { root: true })
      commit(types.SET_ENTITIES, [])
      commit(types.SET_ENTITY_COUNT, -1)
      commit(types.SET_ENTITY_TYPE, undefined)
      return error
    }
  },

  async fetchEntity ({ commit, dispatch }, params: PaperEntityQueryParameters): Promise<void> {
    try {
      const EntityClass = entityKeysMap[params.entityType]
      const response = await Axios(`/projects/${params.projectId}/paper-entities`, {
        method: 'POST',
        data: {
          query: EntityClass.getQuery(params)
        }
      })
      const data = response.data.data[EntityClass.queryName]
      const payload: PaperEntity = EntityClass.transformer.fetch(data[EntityClass.schemaName])
      commit(types.SET_ACTIVE_ENTITY, payload)
      return undefined
    } catch (error) {
      logger.error(error)
      dispatch('toasts/showError', 'project.explore.fetch_single_error', { root: true })
      commit(types.SET_ACTIVE_ENTITY, undefined)
      return error
    }
  },

  async updateEntity ({ commit, dispatch }, {
    params,
    data
  }: { params: PaperEntityMutationParameters, data: Partial<PaperEntityFields>}): Promise<void> {
    try {
      const EntityClass = entityKeysMap[params.entityType]
      const response = await Axios(`/projects/${params.projectId}/paper-entities`, {
        method: 'POST',
        data: {
          mutation: EntityClass.getMutation(params, EntityClass.transformer.send(data))
        }
      })
      const responseData = response.data.data[EntityClass.queryName]
      const payload: PaperEntity = EntityClass.transformer.fetch(responseData)
      commit(types.UPDATE_ENTITY, payload)
      return undefined
    } catch (error) {
      logger.error(error)
      dispatch('toasts/showError', 'project.explore.update_single_error', { root: true })
      return error
    }
  }
}
