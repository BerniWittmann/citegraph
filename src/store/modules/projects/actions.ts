import Axios from 'axios'
import { ActionTree } from 'vuex'
import { ProjectsState } from '@/store/modules/projects/types'
import { RootState } from '@/store/types'
import Project from '@/models/project'
import * as types from './mutation-types'

export const actions: ActionTree<ProjectsState, RootState> = {
  async fetchProject ({ commit }, id: string): Promise<undefined> {
    try {
      const response = await Axios(`/projects/${id}`)
      const payload: Project = new Project(response && response.data)
      commit(types.SET_ACTIVE_PROJECT, payload)
      commit(types.OPEN_PROJECT, payload)
      return undefined
    } catch (error) {
      console.error(error)
      commit(types.UNSET_ACTIVE_PROJECT)
      return error
    }
  },

  unsetActiveProject ({ commit }): void {
    commit(types.UNSET_ACTIVE_PROJECT)
  },

  async fetchProjects ({ commit }): Promise<undefined> {
    try {
      const response = await Axios(`/projects`)
      const projects = response.data.map((project: Project) => new Project(project))
      commit(types.SET_PROJECTS, projects)
      return undefined
    } catch (error) {
      console.error(error)
      return error
    }
  },

  openProject ({ commit }, project: Project): void {
    commit(types.OPEN_PROJECT, project)
  },

  closeProject ({ commit }, project: Project): void {
    commit(types.CLOSE_PROJECT, project)
  },

  async createProject ({ commit }, data: Project): Promise<Project | undefined> {
    try {
      const response = await Axios(`/projects`, {
        method: 'POST',
        data: data
      })
      const project = new Project(response.data)
      commit(types.ADD_PROJECT, project)
      return project
    } catch (error) {
      console.error(error)
      return error
    }
  }
}
