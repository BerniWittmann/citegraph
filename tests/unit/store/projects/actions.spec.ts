import moxios from 'moxios'
import { actions } from '@/store/modules/projects/actions'
import * as mutationTypes from '@/store/modules/projects/mutation-types'
import Project from '@/models/project'

describe('store/modules/projects/actions', () => {
  it('provides the actions', () => {
    expect(actions).toMatchSnapshot()
  })

  describe('unsetActiveProject', () => {
    it('triggers the mutation', () => {
      const commit = jest.fn()
      const action = actions.unsetActiveProject as Function
      action({ commit })
      expect(commit).toHaveBeenCalledWith(mutationTypes.UNSET_ACTIVE_PROJECT)
    })
  })

  describe('fetchProject', () => {
    const projectData = {
      id: 42,
      name: 'My awesome Project'
    }

    beforeEach(() => {
      moxios.install()
    })
    afterEach(() => {
      moxios.uninstall()
    })
    it('fetches the project', (done) => {
      moxios.stubRequest('/projects/42', {
        status: 200,
        response: projectData
      })

      const onFulfilled = jest.fn()
      const commit = jest.fn()
      const action = actions.fetchProject as Function

      action({ commit }, 42).then(onFulfilled)

      moxios.wait(() => {
        expect(onFulfilled).toHaveBeenCalled()
        expect(commit).toHaveBeenCalledWith(mutationTypes.SET_ACTIVE_PROJECT, new Project(projectData))
        expect(commit).toHaveBeenCalledWith(mutationTypes.OPEN_PROJECT, new Project(projectData))
        done()
      })
    })

    it('handles the error when one appeared during the fetching', (done) => {
      moxios.stubRequest('/projects/42', {
        status: 404,
        response: 'The project could not be found'
      })

      const onFulfilled = jest.fn()
      const commit = jest.fn()
      const dispatch = jest.fn()
      console.error = jest.fn()
      const action = actions.fetchProject as Function

      action({ commit, dispatch }, 42).then(onFulfilled)

      moxios.wait(() => {
        expect(onFulfilled).toHaveBeenCalledWith(new Error('Request failed with status code 404'))
        expect(commit).toHaveBeenCalledWith(mutationTypes.UNSET_ACTIVE_PROJECT)
        expect(dispatch).toHaveBeenCalledWith('toasts/showError', 'project.fetch_error', { root: true })
        done()
      })
    })
  })

  describe('fetchProjects', () => {
    const projectsData = [
      { id: 42, name: 'My awesome Project' },
      { id: 1, name: 'First Project' },
      { id: 99, name: 'Last Project' }
    ]

    beforeEach(() => {
      moxios.install()
    })
    afterEach(() => {
      moxios.uninstall()
    })
    it('fetches all projects', (done) => {
      moxios.stubRequest('/projects', {
        status: 200,
        response: projectsData
      })

      const onFulfilled = jest.fn()
      const commit = jest.fn()
      const action = actions.fetchProjects as Function

      action({ commit }, 42).then(onFulfilled)

      moxios.wait(() => {
        expect(onFulfilled).toHaveBeenCalled()
        expect(commit).toHaveBeenCalledWith(mutationTypes.SET_PROJECTS, projectsData.map(project => new Project(project)))
        done()
      })
    })

    it('handles the error when one appeared during the fetching', (done) => {
      moxios.stubRequest('/projects', {
        status: 500,
        response: 'An error occured'
      })

      const onFulfilled = jest.fn()
      const commit = jest.fn()
      const dispatch = jest.fn()
      console.error = jest.fn()
      const action = actions.fetchProjects as Function

      action({ commit, dispatch }, 42).then(onFulfilled)

      moxios.wait(() => {
        expect(onFulfilled).toHaveBeenCalledWith(new Error('Request failed with status code 500'))
        expect(commit).not.toHaveBeenCalled()
        expect(dispatch).toHaveBeenCalledWith('toasts/showError', 'projects.fetch_error', { root: true })
        done()
      })
    })
  })

  describe('openProject', () => {
    it('triggers the mutation', () => {
      const commit = jest.fn()
      const action = actions.openProject as Function
      const project = new Project({ id: 42, name: 'My awesome Project' })
      action({ commit }, project)
      expect(commit).toHaveBeenCalledWith(mutationTypes.OPEN_PROJECT, project)
    })
  })

  describe('closeProject', () => {
    it('triggers the mutation', () => {
      const commit = jest.fn()
      const action = actions.closeProject as Function
      const project = new Project({ id: 42, name: 'My awesome Project' })
      action({ commit }, project)
      expect(commit).toHaveBeenCalledWith(mutationTypes.CLOSE_PROJECT, project)
    })
  })

  describe('createProject', () => {
    const projectData = {
      id: 42,
      name: 'My awesome Project'
    }

    beforeEach(() => {
      moxios.install()
    })
    afterEach(() => {
      moxios.uninstall()
    })
    it('creates a  new project', (done) => {
      moxios.stubRequest('/projects', {
        status: 201,
        response: projectData
      })

      const onFulfilled = jest.fn()
      const commit = jest.fn()
      const dispatch = jest.fn()
      const action = actions.createProject as Function

      action({ commit, dispatch }, projectData).then(onFulfilled)

      moxios.wait(() => {
        expect(onFulfilled).toHaveBeenCalled()
        expect(commit).toHaveBeenCalledWith(mutationTypes.ADD_PROJECT, new Project(projectData))
        expect(onFulfilled).toHaveBeenCalledWith(new Project(projectData))
        expect(dispatch).toHaveBeenCalledWith('toasts/showSuccess', 'projects.add.successful', { root: true })
        done()
      })
    })

    it('handles the error when one appeared during the creation', (done) => {
      moxios.stubRequest('/projects', {
        status: 400,
        response: 'The given name is invalid'
      })

      const onFulfilled = jest.fn()
      const commit = jest.fn()
      const dispatch = jest.fn()
      console.error = jest.fn()
      const action = actions.createProject as Function

      action({ commit, dispatch }, projectData).then(onFulfilled)

      moxios.wait(() => {
        expect(onFulfilled).toHaveBeenCalledWith(new Error('Request failed with status code 400'))
        expect(commit).not.toHaveBeenCalled()
        expect(dispatch).toHaveBeenCalledWith('toasts/showError', 'projects.add.error', { root: true })
        done()
      })
    })
  })

  describe('deleteProject', () => {
    const projectData = {
      id: 42,
      name: 'My awesome Project'
    }

    beforeEach(() => {
      moxios.install()
    })
    afterEach(() => {
      moxios.uninstall()
    })
    it('deletes the project', (done) => {
      moxios.stubRequest('/projects/42', {
        status: 200
      })

      const onFulfilled = jest.fn()
      const commit = jest.fn()
      const dispatch = jest.fn()
      const action = actions.deleteProject as Function

      action({ commit, dispatch }, projectData).then(onFulfilled)

      moxios.wait(() => {
        expect(onFulfilled).toHaveBeenCalled()
        expect(commit).toHaveBeenCalledWith(mutationTypes.CLOSE_PROJECT, new Project(projectData))
        expect(commit).toHaveBeenCalledWith(mutationTypes.DELETE_PROJECT, new Project(projectData))
        expect(dispatch).toHaveBeenCalledWith('toasts/showSuccess', 'project.delete.successful', { root: true })
        done()
      })
    })

    it('handles the error when one appeared during the deletion', (done) => {
      moxios.stubRequest('/projects/42', {
        status: 404,
        response: 'The project could not be found'
      })

      const onFulfilled = jest.fn()
      const commit = jest.fn()
      const dispatch = jest.fn()
      console.error = jest.fn()
      const action = actions.deleteProject as Function

      action({ commit, dispatch }, projectData).then(onFulfilled)

      moxios.wait(() => {
        expect(onFulfilled).toHaveBeenCalledWith(new Error('Request failed with status code 404'))
        expect(commit).not.toHaveBeenCalledWith(mutationTypes.CLOSE_PROJECT, new Project(projectData))
        expect(commit).not.toHaveBeenCalledWith(mutationTypes.DELETE_PROJECT, new Project(projectData))
        expect(dispatch).toHaveBeenCalledWith('toasts/showError', 'project.delete.error', { root: true })
        done()
      })
    })
  })

  describe('updateProject', () => {
    const projectData = {
      id: 42,
      name: 'My awesome Project'
    }

    beforeEach(() => {
      moxios.install()
    })
    afterEach(() => {
      moxios.uninstall()
    })
    it('updates the project', (done) => {
      moxios.stubRequest('/projects/42', {
        status: 200,
        response: projectData
      })

      const onFulfilled = jest.fn()
      const commit = jest.fn()
      const dispatch = jest.fn()
      const action = actions.updateProject as Function

      action({ commit, dispatch }, projectData).then(onFulfilled)

      moxios.wait(() => {
        expect(onFulfilled).toHaveBeenCalled()
        expect(commit).toHaveBeenCalledWith(mutationTypes.UPDATE_PROJECT, new Project(projectData))
        expect(dispatch).toHaveBeenCalledWith('toasts/showSuccess', 'project.edit.successful', { root: true })
        done()
      })
    })

    it('handles the error when one appeared during the update', (done) => {
      moxios.stubRequest('/projects/42', {
        status: 404,
        response: 'The project could not be found'
      })

      const onFulfilled = jest.fn()
      const commit = jest.fn()
      const dispatch = jest.fn()
      console.error = jest.fn()
      const action = actions.updateProject as Function

      action({ commit, dispatch }, projectData).then(onFulfilled)

      moxios.wait(() => {
        expect(onFulfilled).toHaveBeenCalledWith(new Error('Request failed with status code 404'))
        expect(commit).not.toHaveBeenCalledWith(mutationTypes.UPDATE_PROJECT, new Project(projectData))
        expect(dispatch).toHaveBeenCalledWith('toasts/showError', 'project.edit.error', { root: true })
        done()
      })
    })
  })
})
