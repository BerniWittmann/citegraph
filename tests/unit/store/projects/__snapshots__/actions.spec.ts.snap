// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store/modules/projects/actions provides the actions 1`] = `
Object {
  "closeProject": [Function],
  "createProject": [Function],
  "fetchProject": [Function],
  "fetchProjects": [Function],
  "openProject": [Function],
  "unsetActiveProject": [Function],
}
`;
